logger.js

```diff
@@ -10,135 +10,142 @@
  * LICENSE file.
  *
  * Removal or modification of this copyright notice is prohibited.
+ *
  */
+const os = require('os');
+const winston = require('winston');
 
-'use strict';
-
-var fs = require('fs');
-var util = require('util');
-var child_process = require('child_process');
-var path = require('path');
-var strftime = require('strftime').utc();
-
-require('colors');
-
-module.exports = function(config) {
-	config = config || {};
-	var exports = {};
+const { format } = winston;
 
-	config.levels = config.levels || {
-		none: 99,
-		trace: 0,
-		debug: 1,
-		log: 2,
+const customLevels = {
+	levels: {
+		fatal: 0,
+		error: 1,
+		warn: 2,
 		info: 3,
-		warn: 4,
-		error: 5,
-		fatal: 6,
-	};
-
-	config.level_abbr = config.level_abbr || {
-		trace: 'trc',
-		debug: 'dbg',
-		log: 'log',
-		info: 'inf',
-		warn: 'WRN',
-		error: 'ERR',
-		fatal: 'FTL',
-	};
-
-	config.filename = `${process.cwd()}/${config.filename || 'logs.log'}`;
-
-	config.errorLevel = config.errorLevel || 'log';
-
-	child_process.execSync(`mkdir -p ${path.dirname(config.filename)}`);
-	var log_file = fs.createWriteStream(config.filename, { flags: 'a' });
+		debug: 4,
+		query: 5,
+		trace: 10,
+		none: 99,
+	},
+	colors: {
+		fatal: 'yellow',
+		error: 'red',
+		warn: 'meganta',
+		info: 'blue',
+		debug: 'green',
+		query: 'green',
+		trace: 'cyan',
+		none: 'black',
+	},
+};
 
-	exports.setLevel = function(errorLevel) {
-		config.errorLevel = errorLevel;
-	};
+const logRootKeys = [
+	'level',
+	'timestamp',
+	'module',
+	'message',
+	'action',
+	'hostname',
+	'durationMs',
+];
+
+winston.addColors(customLevels.colors);
+
+const defaultLoggerConfig = {
+	filename: 'logs/lisk.log',
+	level: 'none',
+	consoleLevel: 'debug',
+};
 
-	function snipFragileData(data) {
-		for (var key in data) {
-			if (key.search(/passphrase|password/i) > -1) {
-				data[key] = 'XXXXXXXXXX';
+const formatInfo = format(info =>
+	Object.keys(info).reduce((accumulated, key) => {
+		if (logRootKeys.includes(key)) {
+			accumulated[key] = info[key];
+		} else {
+			if (!accumulated.meta) {
+				accumulated.meta = {};
 			}
+			accumulated.meta[key] = info[key];
 		}
-		return data;
-	}
-
-	Object.keys(config.levels).forEach(name => {
-		function log(message, data) {
-			var log = {
-				level: name,
-				timestamp: strftime('%F %T', new Date()),
-			};
+		return accumulated;
+	}, {}));
```

Comment:
We have to revisit the whole logic here. In one single operations we may be logging 20-30 log entries. For each log entry:

- This function is looping over all keys in json object.
- Then for each key its looping over another array `logRootKeys` to validate `includes`.
- So if a log object contains 8 keys, then logging that object may need 56 loop iterations and 8 function calls. 
- Also initializing default value `.meta = {}` does not make sense here, instead should be done on log object initialization. 

Overall I believe this logging approach will create a lot of unwanted overhead to system performance.

Reply:
Yes main reason of this is more like a util function to format the meta data, so we can use it like
`logger.info('some message', { thiswill: 'be in meta'});`

if we want to write it like
`logger.info('some message', { meta: { thiswill: 'be in meta'}});` we don't need this process